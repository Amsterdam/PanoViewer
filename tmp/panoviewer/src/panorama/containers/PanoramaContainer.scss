@import "../../shared/styles/config";

.c-panorama {
  bottom: 0;
  left: 0;
  position: absolute;
  right: 0;
  top: 1px;
  z-index: 1;
}

.c-panorama__marzipano {
  border: solid $secondary-gray20 1px;
  bottom: 0;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
}

.c-panorama__status-bar {
  bottom: 0;
  left: 0;
  position: absolute;
  right: 0;
  z-index: 2;
}

.c-panorama__close {
  @extend %hover-focus-button;
  position: absolute;
  right: $container-whitespace;
  top: $container-whitespace;
  z-index: 2;
}

.is-print-mode,
.is-embed-preview,
.is-embed {
  .c-panorama {
    bottom: auto;
    height: 800px;
    left: auto;
    margin-top: $base-whitespace * 2;
    position: relative;
    right: auto;
    top: auto;
    width: 100%;
  }

  .c-panorama__close {
    display: none;
  }

  .c-panorama__status-bar {
    background-color: $primary-light;
  }
}
