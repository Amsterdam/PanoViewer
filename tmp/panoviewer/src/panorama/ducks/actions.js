// Actions creators
import {
  CLOSE_PANORAMA,
  FETCH_PANORAMA_ERROR,
  FETCH_PANORAMA_HOTSPOT_REQUEST,
  FETCH_PANORAMA_REQUEST,
  FETCH_PANORAMA_REQUEST_EXTERNAL,
  FETCH_PANORAMA_SUCCESS,
  SET_PANORAMA_LOCATION,
  SET_PANORAMA_ORIENTATION,
  SET_PANORAMA_TAGS
} from './constants';

export const fetchPanoramaRequest = (payload) => ({
  type: FETCH_PANORAMA_REQUEST,
  payload,
  meta: {
    tracking: true
  }
});

export const fetchPanoramaHotspotRequest = (payload) => ({
  type: FETCH_PANORAMA_HOTSPOT_REQUEST,
  payload,
  meta: {
    tracking: true
  }
});

export const setPanoramaTags = (payload) => ({
  type: SET_PANORAMA_TAGS,
  payload,
  meta: {
    tracking: payload
  }
});

export const fetchPanoramaRequestExternal = () => ({
  type: FETCH_PANORAMA_REQUEST_EXTERNAL,
  meta: {
    tracking: true
  }
});

export const fetchPanoramaSuccess = (payload) => ({
  type: FETCH_PANORAMA_SUCCESS,
  payload,
  meta: {
    tracking: payload
  }
});
export const fetchPanoramaError = (error) => ({
  type: FETCH_PANORAMA_ERROR,
  payload: error
});

export const closePanorama = () => ({
  type: CLOSE_PANORAMA
});
export const setPanoramaLocation = (payload) => ({
  type: SET_PANORAMA_LOCATION,
  payload,
  meta: {
    tracking: true
  }
});
export const setPanoramaOrientation = ({ heading, pitch, fov }) => ({
  type: SET_PANORAMA_ORIENTATION,
  payload: {
    heading,
    pitch,
    fov
  }
});
