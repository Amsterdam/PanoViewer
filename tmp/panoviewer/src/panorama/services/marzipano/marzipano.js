import Marzipano from 'marzipano';
import React from 'react';
import { renderToStaticMarkup } from 'react-dom/server';
import { PANORAMA_CONFIG } from '../panorama-api/panorama-api';
import { degreesToRadians, radiansToDegrees } from '../../../shared/services/angle-conversion/angle-conversion';
import Hotspot from '../../components/Hotspot/Hotspot';

/*
  Calculate height of the hotspots
  TODO: DP-6135 replace with height returned from the api
*/
function calculateHotspotPitch(height, distance) {
  return Math.atan(height / distance);
}

/*
  Create hotspot element that will be placed in the hotspot container
*/
function createHotspotTemplate(viewer, scene, view, hotspot, onClickHandler) {
  const hotspotPitch = calculateHotspotPitch(PANORAMA_CONFIG.CAMERA_HEIGHT, hotspot.distance);

  const realLifeHotspotSize = 0.6;
  const minDistance = 4;
  const maxDistance = 21;
  const viewport = 960;

  /*
    All hotspots are shown, the min- and maxDistance variables are only used to determine
    the minimum and maximum hotspot size.
  */
  const correctedDistance = Math.min(maxDistance, Math.max(minDistance, hotspot.distance));
  const viewAngle = Math.atan(realLifeHotspotSize / correctedDistance);

  /*
    The actual hotspot size is dependent on the width of the straatbeeld and the FOV. For
    this first version we're making assumptions about the viewport and FOV.
    Offset is a value between 7 and 10 degrees depending on the distance. it is subtracted from
    the angle of the hotspot x rotation to render hotspots better that are far away
  */
  const offset = (5 / (maxDistance - correctedDistance + 1)) + 8; // eslint-disable-line
  const angle = (90 - radiansToDegrees(hotspotPitch) - offset); // eslint-disable-line

  const size = Math.round((radiansToDegrees(viewAngle) * viewport) / PANORAMA_CONFIG.DEFAULT_FOV);

  /*
    Create hotspot element and render to static markup ReactdomElement
  */
  const hotspotElement = document.createElement('div');
  hotspotElement.className = 'hotspot';
  hotspotElement.innerHTML = renderToStaticMarkup( // eslint-disable-next-line
    <Hotspot
      year={hotspot.year}
      size={size}
      angle={angle}
    />
  );

  // Add onClick event to hotspot
  hotspotElement.addEventListener('click', () => {
    onClickHandler(hotspot.id, hotspot.heading);
  });

  const position = {
    yaw: degreesToRadians(hotspot.heading),
    pitch: hotspotPitch
  };

  return scene.hotspotContainer().createHotspot(hotspotElement, position);
}

/*
  Set the initial Marzipano viewer that is tied to the ref in the PanoramaContainer
*/
export function initialize(domElement) {
  const viewer = new Marzipano.Viewer(domElement, {
    stageType: null,
    stage: {
      preserveDrawingBuffer: true,
      width: 960
    }
  });

  return viewer;
}

/*
  Load the Marzipano viewer into its view components
*/
export function loadScene(viewer, onClickHandler, image, heading, pitch, fov, hotspots) {
  const source = Marzipano.ImageUrlSource.fromString(
    image.pattern,
    { cubeMapPreviewUrl: image.preview }
  );

  const viewLimiter = Marzipano.RectilinearView.limit.traditional(
    PANORAMA_CONFIG.MAX_RESOLUTION,
    degreesToRadians(PANORAMA_CONFIG.MAX_FOV)
  );

  const view = new Marzipano.RectilinearView(null, viewLimiter);

  const scene = viewer.createScene({
    source,
    geometry: new Marzipano.CubeGeometry(PANORAMA_CONFIG.LEVEL_PROPERTIES_LIST),
    view,
    pinFirstLevel: true
  });

   // Do not mutate someone else's data collection!
  const hotspotsObject = JSON.parse(JSON.stringify(hotspots));

  hotspotsObject
    .sort((hotspotA, hotspotB) => hotspotB.distance - hotspotA.distance)
    .forEach((hotspot) => createHotspotTemplate(viewer, scene, view, hotspot, onClickHandler));

  view.setYaw(degreesToRadians(heading));
  view.setPitch(degreesToRadians(pitch));
  view.setFov(degreesToRadians(fov));

  return scene.switchTo();
}

/*
  Return the orientation values according to the view in Marzipano
*/
export function getOrientation(viewer) {
  const heading = radiansToDegrees(viewer.view().yaw());
  const pitch = radiansToDegrees(viewer.view().pitch());
  const fov = radiansToDegrees(viewer.view().fov());

  return {
    heading,
    pitch,
    fov
  };
}

export function getHeadingDegrees([x1, y1], [x2, y2]) {
  return (Math.atan2(y2 - y1, x2 - x1) * 180) / Math.PI;
}
