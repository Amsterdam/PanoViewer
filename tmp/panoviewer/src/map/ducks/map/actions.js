import { getLabelObjectByTags } from '../../../panorama/ducks/selectors';
import { normalizeCoordinate } from '../../../shared/services/coordinate-reference-system';
import {
  MAP_EMPTY_GEOMETRY, MAP_UPDATE_SHAPE, MAP_SET_DRAWING_MODE, MAP_END_DRAWING,
  MAP_CLEAR, MAP_ZOOM, TOGGLE_MAP_PANEL, CLOSE_MAP_PANEL, SET_MAP_BASE_LAYER,
  TOGGLE_MAP_OVERLAY, TOGGLE_MAP_OVERLAY_PANORAMA, TOGGLE_MAP_OVERLAY_VISIBILITY,
  MAP_PAN, SET_MAP_CLICK_LOCATION, MAP_BOUNDING_BOX, MAP_LOADING
} from './constants';

// Actions
export const mapEmptyGeometry = () => ({ type: MAP_EMPTY_GEOMETRY });
export const mapUpdateShape = (payload) => ({ type: MAP_UPDATE_SHAPE, payload });
export const mapSetDrawingMode = (payload) => ({
  type: MAP_SET_DRAWING_MODE,
  payload,
  meta: {
    tracking: payload.drawingMode
  }
});
export const mapEndDrawing = (payload) => ({ type: MAP_END_DRAWING, payload });
export const mapClear = () => ({ type: MAP_CLEAR });
export const updateZoom = (payload) => ({ type: MAP_ZOOM, payload });
export const toggleMapPanel = () => ({ type: TOGGLE_MAP_PANEL });
export const closeMapPanel = () => ({ type: CLOSE_MAP_PANEL });
export const setMapBaseLayer = (payload) => ({
  type: SET_MAP_BASE_LAYER,
  payload,
  meta: {
    tracking: payload
  }
});
export const toggleMapOverlay = (payload) => ({
  type: TOGGLE_MAP_OVERLAY,
  payload: {
    mapLayers: (payload.id) ? [payload.id] : payload.legendItems.map((overlay) => overlay.id)
  },
  meta: {
    tracking: payload
  }
});

export const toggleMapOverlayPanorama = (tags) => ({
  type: TOGGLE_MAP_OVERLAY_PANORAMA,
  payload: getLabelObjectByTags(tags).layerId
});

export const toggleMapOverlayVisibility = (mapLayerId, isVisible) => ({
  type: TOGGLE_MAP_OVERLAY_VISIBILITY,
  mapLayerId,
  isVisible: !isVisible
});
export const updatePan = (payload) => ({
  type: MAP_PAN,
  payload: {
    latitude: normalizeCoordinate(payload.lat, 7),
    longitude: normalizeCoordinate(payload.lng, 7)
  }
});
export const setSelectedLocation = (payload) => ({
  type: SET_MAP_CLICK_LOCATION,
  payload: {
    location: {
      latitude: normalizeCoordinate(payload.latlng.lat, 7),
      longitude: normalizeCoordinate(payload.latlng.lng, 7)
    }
  },
  meta: {
    tracking: true
  }
});
export const updateBoundingBox = (payload) => ({
  type: MAP_BOUNDING_BOX,
  payload
});

export const mapLoadingAction = (payload) => ({ type: MAP_LOADING, payload });
