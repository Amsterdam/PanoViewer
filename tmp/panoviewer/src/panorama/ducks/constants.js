import { toHome } from '../../store/redux-first-router/actions';

export const REDUCER_KEY = 'panorama';
export const FETCH_PANORAMA_REQUEST = `${REDUCER_KEY}/FETCH_PANORAMA_REQUEST`;
export const FETCH_PANORAMA_SUCCESS = `${REDUCER_KEY}/FETCH_PANORAMA_SUCCESS`;
export const FETCH_PANORAMA_ERROR = `${REDUCER_KEY}/FETCH_PANORAMA_ERROR`;
export const SET_PANORAMA_ORIENTATION = `${REDUCER_KEY}/SET_PANORAMA_ORIENTATION`;
export const SET_PANORAMA_LOCATION = `${REDUCER_KEY}/SET_PANORAMA_LOCATION`;
export const CLOSE_PANORAMA = `${REDUCER_KEY}/CLOSE_PANORAMA`;
export const FETCH_PANORAMA_HOTSPOT_REQUEST = `${REDUCER_KEY}/FETCH_PANORAMA_HOTSPOT_REQUEST`;
export const SET_PANORAMA_TAGS = `${REDUCER_KEY}/FETCH_PANORAMA_REQUEST_TOGGLE`;
export const FETCH_PANORAMA_REQUEST_EXTERNAL = `${REDUCER_KEY}/FETCH_PANORAMA_REQUEST_EXTERNAL`;

export const PANO_LABELS = [
  {
    layerId: 'pano',
    label: 'Meest recent',
    tags: ['mission-bi']
  },
  {
    layerId: 'pano2018bi',
    label: 'Alleen 2018',
    tags: ['mission-bi', 'mission-2018']
  },
  {
    layerId: 'pano2017bi',
    label: 'Alleen 2017',
    tags: ['mission-bi', 'mission-2017']
  },
  {
    layerId: 'pano2016bi',
    label: 'Alleen 2016',
    tags: ['mission-bi', 'mission-2016']
  },
  {
    layerId: 'pano2018woz',
    label: 'Alleen 2018 WOZ',
    tags: ['mission-woz', 'mission-2018']
  },
  {
    layerId: 'pano2017woz',
    label: 'Alleen 2017 WOZ',
    tags: ['mission-woz', 'mission-2017']
  }
];

export const initialState = {
  location: null,   // eg: [52.8, 4.9]
  tags: PANO_LABELS[0].tags,
  targetLocation: null,
  pitch: 0,         // eg: -10
  heading: 0,       // eg: 270
  fov: null,        // eg: 65
  image: null,      // eg: {
                    //     pattern: 'http://www.example.com/path/some-id/{this}/{that}/{thingie}.jpg',
                    //     preview: 'http://www.example.com/path/some-id/preview.jpg'
                    // }
  hotspots: [],     // eg: [{id: 'ABC124', heading: 90, distance: 18}],
  date: null,       // eg: new Date()
  isLoading: true,
  detailReference: [],
  pageReference: ''
};

export const PAGE_REFS = {
  HOME: 'home'
};

export const PAGE_REF_MAPPING = {
  [PAGE_REFS.HOME]: toHome
};
