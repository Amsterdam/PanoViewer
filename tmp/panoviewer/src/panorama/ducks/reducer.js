import { PANORAMA_CONFIG } from '../services/panorama-api/panorama-api';
import PAGES from '../../app/pages';
import {
  FETCH_PANORAMA_ERROR,
  FETCH_PANORAMA_HOTSPOT_REQUEST,
  FETCH_PANORAMA_REQUEST,
  FETCH_PANORAMA_SUCCESS,
  initialState,
  REDUCER_KEY,
  SET_PANORAMA_LOCATION,
  SET_PANORAMA_ORIENTATION,
  SET_PANORAMA_TAGS
} from './constants';
import paramsRegistry from '../../store/params-registry';
import { shouldResetState } from '../../store/redux-first-router/actions';

export { REDUCER_KEY as PANORAMA };

export default function reducer(state = initialState, action) {
  if (shouldResetState(action, [PAGES.PANORAMA])) {
    return initialState;
  }

  const enrichedState = {
    ...state,
    ...paramsRegistry.getStateFromQueries(REDUCER_KEY, action)
  };

  switch (action.type) {
    case FETCH_PANORAMA_HOTSPOT_REQUEST:
    case FETCH_PANORAMA_REQUEST:
      return {
        ...enrichedState,
        isLoading: true,
        targetLocation: null
      };

    case SET_PANORAMA_TAGS:
      return {
        ...enrichedState,
        tags: action.payload
      };

    case FETCH_PANORAMA_SUCCESS:
      return {
        ...enrichedState,
        id: action.payload.id,
        date: action.payload.date,
        pitch: enrichedState.pitch || initialState.pitch,
        fov: enrichedState.fov || PANORAMA_CONFIG.DEFAULT_FOV,
        hotspots: action.payload.hotspots,
        isLoading: false,
        location: action.payload.location,
        image: action.payload.image
      };
    case FETCH_PANORAMA_ERROR:
      return {
        ...enrichedState,
        isLoading: false,
        isError: true // TODO: refactor, show error
      };

    case SET_PANORAMA_ORIENTATION:
      return {
        ...enrichedState,
        heading: action.payload.heading,
        pitch: action.payload.pitch,
        fov: action.payload.fov
      };

    case SET_PANORAMA_LOCATION:
      return {
        ...enrichedState,
        location: action.payload,
        targetLocation: action.payload,
        detailReference: initialState.detailReference
      };

    default:
      return enrichedState;
  }
}

