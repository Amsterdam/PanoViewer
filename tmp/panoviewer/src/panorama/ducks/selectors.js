import { createSelector } from 'reselect';
import {
  panoramaOrientationType,
  panoramaPersonType
} from '../../map/components/leaflet/services/icons.constant';
import { PANO_LABELS, REDUCER_KEY } from './constants';

export const getPanorama = (state) => state[REDUCER_KEY] || {};
export const getPanoramaLocation = createSelector(
  getPanorama,
  (panorama) => (
    panorama ? panorama.location : []
  )
);
const getStateOfKey = (key) => (state) => createSelector(getPanorama, (data) => (data[key]))(state);

export const getPanoramaHeading = getStateOfKey('heading');
export const getPanoramaMarkers = createSelector([getPanoramaLocation, getPanoramaHeading],
  (location, heading) => (
    location ? [
      {
        position: location,
        type: panoramaOrientationType,
        heading: heading || 0
      },
      {
        position: location,
        type: panoramaPersonType
      }
    ] : []
  )
);

export const getPanoramaPitch = getStateOfKey('pitch');
export const getDetailReference = getStateOfKey('detailReference');
export const getPageReference = getStateOfKey('pageReference');
export const getPanoramaTags = getStateOfKey('tags');

export const getLabelObjectByTags = (tags) =>
  PANO_LABELS.find((value) => (
    JSON.stringify(value.tags.sort()) === JSON.stringify(tags)
  )) || PANO_LABELS[0];
