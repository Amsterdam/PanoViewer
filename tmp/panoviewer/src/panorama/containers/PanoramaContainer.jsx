import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import throttle from 'lodash.throttle';

import './PanoramaContainer.scss';
import {
  closePanorama,
  fetchPanoramaHotspotRequest,
  setPanoramaOrientation
} from '../ducks/actions';

import {
  getHeadingDegrees,
  getOrientation,
  initialize,
  loadScene
} from '../services/marzipano/marzipano';

import StatusBar from '../components/StatusBar/StatusBar';
import ToggleFullscreen from '../../app/components/ToggleFullscreen/ToggleFullscreen';
import {
  getDetailReference,
  getLabelObjectByTags,
  getPanorama,
  getPanoramaLocation,
  getPanoramaTags
} from '../ducks/selectors';
import IconButton from '../../app/components/IconButton/IconButton';
import { getMapDetail } from '../../map/ducks/detail/map-detail';
import { getMapOverlays } from '../../map/ducks/map/selectors';
import { pageTypeToEndpoint } from '../../map/services/map-detail';
import { setViewMode, VIEW_MODE } from '../../shared/ducks/ui/ui';

class PanoramaContainer extends React.Component {
  constructor(props) {
    super(props);
    this.toggleFullscreen = this.toggleFullscreen.bind(this);
    this.updateOrientation = this.updateOrientation.bind(this);
    this.hotspotClickHandler = this.hotspotClickHandler.bind(this);
    this.loadPanoramaScene = this.loadPanoramaScene.bind(this);

    this.updateOrientationThrottled = throttle(
      this.updateOrientation,
      300,
      {
        leading: true,
        trailing: true
      }
    );
  }

  componentDidMount() {
    const { detailReference, fetchMapDetail } = this.props;
    this.panoramaViewer = initialize(this.panoramaRef);
    this.loadPanoramaScene();

    if (this.panoramaViewer) {
      this.panoramaViewer.addEventListener('viewChange', this.updateOrientationThrottled);
    }

    if (detailReference.length > 0) {
      const [id, type, subtype] = detailReference;
      fetchMapDetail(pageTypeToEndpoint(type, subtype, id));
    }
  }

  componentDidUpdate(prevProps) {
    const { panoramaState } = this.props;

    if (panoramaState.image !== prevProps.panoramaState.image) {
      this.loadPanoramaScene();
    }
  }

  componentWillUnmount() {
    this.panoramaViewer.removeEventListener('viewChange', this.updateOrientationThrottled);
  }

  loadPanoramaScene() {
    const { panoramaState } = this.props;
    const { heading: currentHeading, location, targetLocation } = panoramaState;
    const heading = (Array.isArray(location) && Array.isArray(targetLocation))
      ? getHeadingDegrees(location, targetLocation)
      : currentHeading;
    if (panoramaState.image) {
      loadScene(
        this.panoramaViewer,
        this.hotspotClickHandler,
        panoramaState.image,
        heading,
        panoramaState.pitch,
        panoramaState.fov,
        panoramaState.hotspots
      );
    }
  }

  updateOrientation() {
    const { heading, pitch, fov } = getOrientation(this.panoramaViewer);
    this.props.setOrientation({ heading, pitch, fov });
  }

  hotspotClickHandler(panoramaId) {
    const { fetchPanoramaById } = this.props;
    return fetchPanoramaById({ id: panoramaId });
  }

  toggleFullscreen() {
    const { isFullscreen, setView } = this.props;

    if (isFullscreen) {
      return setView(VIEW_MODE.SPLIT, 'beeld-verkleinen');
    }

    return setView(VIEW_MODE.FULL, 'beeld-vergroten');
  }

  render() {
    const {
      isFullscreen,
      panoramaState,
      onClose,
      tags
    } = this.props;
    return (
      <div className="c-panorama">
        <div
          ref={
            //eslint-disable-next-line
            (el) => this.panoramaRef = el
          }
          role="button"
          tabIndex="-1"
          className="c-panorama__marzipano js-marzipano-viewer"
        />

        <ToggleFullscreen
          isFullscreen={isFullscreen}
          title="Panoramabeeld"
          onToggleFullscreen={this.toggleFullscreen}
          alignLeft
        />

        <IconButton
          onClick={onClose}
          title="Sluit panorama"
          icon="cross"
        />

        {(panoramaState.date && panoramaState.location) ? (
          <StatusBar
            date={panoramaState.date}
            location={panoramaState.location}
            heading={panoramaState.heading}
            currentLabel={getLabelObjectByTags(tags).label}
          />
        ) : ''}
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  panoramaState: getPanorama(state),
  tags: getPanoramaTags(state),
  detailReference: getDetailReference(state),
  pageReference: getDetailReference(state),
  panoramaLocation: getPanoramaLocation(state),
  overlays: getMapOverlays(state)
});

const mapDispatchToProps = (dispatch) => ({
  ...bindActionCreators({
    setOrientation: setPanoramaOrientation,
    setView: setViewMode,
    fetchPanoramaById: fetchPanoramaHotspotRequest,
    fetchMapDetail: getMapDetail,
    onClose: closePanorama
  }, dispatch)
});

PanoramaContainer.propTypes = {
  panoramaState: PropTypes.shape({}).isRequired,
  isFullscreen: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  setView: PropTypes.func.isRequired,
  tags: PropTypes.arrayOf(PropTypes.string).isRequired,
  detailReference: PropTypes.arrayOf(PropTypes.string).isRequired,
  setOrientation: PropTypes.func.isRequired,
  fetchMapDetail: PropTypes.func.isRequired,
  fetchPanoramaById: PropTypes.func.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(PanoramaContainer);
