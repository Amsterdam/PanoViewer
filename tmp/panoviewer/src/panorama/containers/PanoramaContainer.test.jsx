import React from 'react';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import PanoramaContainer from './PanoramaContainer';
import { getOrientation, loadScene } from '../services/marzipano/marzipano';
import { fetchPanoramaHotspotRequest } from '../ducks/actions';
import {
  getDetailReference,
  getLabelObjectByTags,
  getPanorama,
  getPanoramaLocation,
  getPanoramaTags
} from '../ducks/selectors';
import { getMapOverlays } from '../../map/ducks/map/selectors';
import { setViewMode, VIEW_MODE } from '../../shared/ducks/ui/ui';

jest.mock('../../map/ducks/map/selectors');
jest.mock('../services/marzipano/marzipano');
jest.mock('../ducks/selectors');
jest.mock('../../shared/ducks/ui/ui');

describe('PanoramaContainer', () => {
  const initialState = {};
  const store = configureMockStore()({ ...initialState });
  const props = {
    isFullscreen: false,
    detailReference: []
  };

  getPanorama.mockImplementation(() => ({
    id: 'ABC',
    heading: 999,
    image: 'ABC_IMAGE.jpg',
    date: '2012-12-12T00:00:00.000Z',
    location: [1, 2]
  }));
  getLabelObjectByTags.mockImplementation(() => ({ label: 'Meest recent' }));
  getPanoramaTags.mockImplementation(() => (['mission-bi']));
  setViewMode.mockImplementation(() => ({ type: 'some type' }));
  getDetailReference.mockImplementation(() => []);
  getMapOverlays.mockImplementation(() => ([]));
  getPanoramaLocation.mockImplementation(() => []);

  beforeEach(() => {
    jest.spyOn(store, 'dispatch');
  });

  afterEach(() => {
    store.dispatch.mockClear();
  });

  it('should render everything', () => {
    const wrapper = shallow(
      <PanoramaContainer {...props} />, { context: { store } }
    ).dive();

    expect(wrapper).toMatchSnapshot();
  });

  it('should render StatusBar when panoramaState is complete', () => {
    const wrapper = shallow(
      <PanoramaContainer {...props} />, { context: { store } }
    ).dive();

    expect(wrapper).toMatchSnapshot();
  });

  it('should load new scene when panorama image information changes', () => {
    getOrientation.mockReturnValue({ heading: 999, pitch: 10, fov: 80 });
    const wrapper = shallow(
      <PanoramaContainer {...props} />, { context: { store } }
    ).dive();

    wrapper.instance().hotspotClickHandler('XYZ');
    expect(store.dispatch).toHaveBeenCalledWith(fetchPanoramaHotspotRequest({ id: 'XYZ' }));
  });

  it('should toggle size of panorama image', () => {
    jest.spyOn(store, 'dispatch');
    const wrapper = shallow(
      <PanoramaContainer {...props} />, { context: { store } }
    ).dive();

    expect(wrapper.instance().props.isFullscreen).toBe(false);

    wrapper.instance().toggleFullscreen();
    expect(store.dispatch).toHaveBeenCalledWith(setViewMode(VIEW_MODE.FULL));

    wrapper.setProps({ isFullscreen: true });

    wrapper.instance().toggleFullscreen();
    expect(store.dispatch).toHaveBeenCalledWith(setViewMode(VIEW_MODE.SPLIT));
  });

  it('should load new scene when panorama image information changes', () => {
    loadScene.mockImplementation();
    const wrapper = shallow(
      <PanoramaContainer {...props} />, { context: { store } }
    ).dive();

    wrapper.setProps({ panoramaState: { image: 'ABC_IMAGE_2.jpg' } });
    wrapper.instance().setState({ update: true });
    wrapper.update();

    expect(wrapper.instance().props.panoramaState.image).toBe('ABC_IMAGE_2.jpg');
    expect(loadScene).toHaveBeenCalled();
  });
});
