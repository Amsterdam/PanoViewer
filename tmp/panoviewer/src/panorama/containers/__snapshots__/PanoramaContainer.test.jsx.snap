// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PanoramaContainer should render StatusBar when panoramaState is complete 1`] = `
<div
  className="c-panorama"
>
  <div
    className="c-panorama__marzipano js-marzipano-viewer"
    role="button"
    tabIndex="-1"
  />
  <ToggleFullscreen
    alignLeft={true}
    isFullscreen={false}
    onToggleFullscreen={[Function]}
    title="Panoramabeeld"
  />
  <IconButton
    alignLeft={false}
    extraClass=""
    icon="cross"
    onClick={[Function]}
    title="Sluit panorama"
  />
  <StatusBar
    currentLabel="Meest recent"
    date="2012-12-12T00:00:00.000Z"
    heading={999}
    location={
      Array [
        1,
        2,
      ]
    }
  />
</div>
`;

exports[`PanoramaContainer should render everything 1`] = `
<div
  className="c-panorama"
>
  <div
    className="c-panorama__marzipano js-marzipano-viewer"
    role="button"
    tabIndex="-1"
  />
  <ToggleFullscreen
    alignLeft={true}
    isFullscreen={false}
    onToggleFullscreen={[Function]}
    title="Panoramabeeld"
  />
  <IconButton
    alignLeft={false}
    extraClass=""
    icon="cross"
    onClick={[Function]}
    title="Sluit panorama"
  />
  <StatusBar
    currentLabel="Meest recent"
    date="2012-12-12T00:00:00.000Z"
    heading={999}
    location={
      Array [
        1,
        2,
      ]
    }
  />
</div>
`;
